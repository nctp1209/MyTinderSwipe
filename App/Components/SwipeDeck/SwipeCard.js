import React, { Component } from "react";
import {
  StyleSheet,
  Text,
  Animated,
  PanResponder,
  Dimensions
} from "react-native";
import PropTypes from "prop-types";

const SCREEN_WIDTH = Dimensions.get("window").width;
const SCREEN_HEIGHT = Dimensions.get("window").height;

export default class DeckContainer extends Component {
  constructor() {
    super();
    this.position = new Animated.ValueXY();
    this.rotate = this.position.x.interpolate({
      inputRange: [-SCREEN_WIDTH / 2, 0, SCREEN_WIDTH / 2],
      outputRange: ["-10deg", "0deg", "10deg"],
      extrapolate: "clamp"
    });
  }

  static propTypes = {
    index: PropTypes.number,
    onSwipeLeft: PropTypes.func,
    onSwipeRight: PropTypes.func,
    renderCardItem: PropTypes.func
  };

  componentWillMount() {
    this.PanResponder = PanResponder.create({
      onStartShouldSetPanResponder: (evt, gestureState) => true,
      onPanResponderMove: (evt, gestureState) => {
        this.position.setValue({ x: gestureState.dx, y: gestureState.dy });
      },
      onPanResponderRelease: (evt, gestureState) => {
        // Move to my favourite
        if (gestureState.dx > 120) {
          this.props.onSwipeRight();
          Animated.spring(this.position, {
            toValue: { x: SCREEN_WIDTH + 100, y: gestureState.dy }
          }).start(() => {});

          // Remove from deck
        } else if (gestureState.dx < -120) {
          this.props.onSwipeLeft();
          Animated.spring(this.position, {
            toValue: { x: -SCREEN_WIDTH - 100, y: gestureState.dy }
          }).start(() => {});

          // Do nothing
        } else {
          Animated.spring(this.position, {
            toValue: { x: 0, y: 0 },
            friction: 4
          }).start();
        }
      }
    });
  }

  render() {
    return (
      <Animated.View
        {...this.PanResponder.panHandlers}
        style={[
          {
            transform: [
              { rotate: this.rotate },
              ...this.position.getTranslateTransform()
            ]
          },
          styles.cardStyle
        ]}
      >
        {this.props.renderCardItem()}
      </Animated.View>
    );
  }
}

const styles = StyleSheet.create({
  cardStyle: {
    height: SCREEN_HEIGHT - 200,
    width: SCREEN_WIDTH - 40,
    margin: 20,
    borderRadius: 5,
    backgroundColor: "white",
    position: "absolute"
  }
});
