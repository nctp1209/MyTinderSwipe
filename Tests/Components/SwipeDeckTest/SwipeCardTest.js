import React from "react";
import { Text } from "react-native";
import renderer from "react-test-renderer";
import { path } from "ramda";
import { shallow } from "enzyme";

import APIResult from "../../APIResult";
import SwipeCard from "../../../App/Components/SwipeDeck/SwipeCard";

test("SwipeCard render", () => {
  const response = APIResult.getPeople();
  const item = path(["data", "results"], response)[0];
  const tree = renderer
    .create(
      <SwipeCard
        key={0}
        index={0}
        renderCardItem={() => <Text>{item.user.username}</Text>}
        onSwipeLeft={() => console.log("onSwipeLeft")}
        onSwipeRight={() => console.log("onSwipeRight")}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
