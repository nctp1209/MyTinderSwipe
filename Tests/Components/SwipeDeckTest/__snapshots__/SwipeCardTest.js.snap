// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeCard render 1`] = `
<View
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 5,
      "height": 1134,
      "margin": 20,
      "position": "absolute",
      "transform": Array [
        Object {
          "rotate": "0deg",
        },
        Object {
          "translateX": 0,
        },
        Object {
          "translateY": 0,
        },
      ],
      "width": 710,
    }
  }
>
  <Text>
    tinymeercat81
  </Text>
</View>
`;
